### ISSUE TITLE
Add feature for hitscan attacks and missile trajectories are affected by mouse look

### ISSUE THREAD

**Issue by JNechaevsky** [fabiangreffrath/crispy-doom#253]
I have an working solution for this, even with support of your `P_SpawnPuffSafe` implementation, but before we move on further -- you haven't added this feature because of it's unsafe for both demos and network games, right? If so, let's just close this issue. 

I know you are using `singleplayer` only in case of critical necessity, but maybe it's worth to make just a little exception for this feature? It's making mouse look usage much more logical and good looking.

Also, PrBoom+ starting from version 2.5.1.4 also have nominal support for vertical trajectories, which are seems to be safe for compatibility, but frankly, looking kinda bad with hitting floors - bullet puffs just visually "slipping" through the floor to the nearest wall.

**Comment by fabiangreffrath** [comment]
I am not exactly sure what you mean.

It is already possible to enable "direct aiming" i.e. shooting under the same angle that you look into using mouse look. Missiles like e.g. rockets will explode at the correct location when hitting either floors/ceilings or walls, but bullet puffs will only appear on walls, even if hitting a ceiling/floor. 

In the latter case, the trajectory will pass through the ceiling/floor until it hits a wall and then the bullet puff will be spawned there but at the hightest/lowest possible vertical position, i.e. directly under the ceiling or above the floor. 

Are you supposing a fix for this phenomenon?

**Comment by JNechaevsky** [comment]
Oops, sorry, haven't noticed "vertical aiming" menu item!

Yep, I'm suggesting to fix passing through floors/ceilings to achive this: [screenshot](http://screenshot.ru/upload/images/2018/01/15/puff35992.png). Since `PTR_ShootTraverse` is a boolean, no need to change it, better to add another one just like I've did here https://github.com/JNechaevsky/russian-doom/commit/4bad89209716807ad96ef0c4757d107ccb43f404. It's slightly complicated, but working fine.

But beware, despite of cool looking and proper logic, it's unsafe for internal demos, and probably will be very unsafe for network games - some players may have enabled this feature, some are not. In other words: it's totally non-vanilla behavior and probably will not be able to work with same logic as standard `PTR_ShootTraverse`, but still safe for vertical autoaiming. That's why I have defined it strictly as `singleplayer`.

**Comment by fabiangreffrath** [comment]
It should be safe as long as the bullet puff would appear anyway, but now it appears at another position. The whole thing becomes critical if we spawn bullet puffs where they normally wouldn't or - the other way round - don't spawn bullet puffs e.g. because we are shooting at the sky whereas thex would normally appear at the wall surrounding that sky.

**Comment by fabiangreffrath** [comment]
Lighter version of your patch that works properly at least in buildings:
```
--- a/src/doom/p_map.c
+++ b/src/doom/p_map.c
@@ -1080,6 +1080,15 @@ boolean PTR_ShootTraverse (intercept_t* in)
                safe = true;
            }
        }
+       else
+       if (z < li->frontsector->floorheight ||
+           z > li->frontsector->ceilingheight)
+       {
+               z = (z < li->frontsector->floorheight) ? li->frontsector->floorheight : li->frontsector->ceilingheight;
+               frac = FixedDiv(z - shootz, FixedMul(aimslope, attackrange));
+               x = trace.x + FixedMul (trace.dx, frac);
+               y = trace.y + FixedMul (trace.dy, frac);
+       }
 
        // [crispy] update laser spot position and return
        if (la_damage == INT_MIN)

```

**Comment by JNechaevsky** [comment]
Just checked out. It's working in some instances, but in some are not:

- Working: http://screenshot.ru/upload/images/2018/01/16/11a2b8.png
- Not working: http://screenshot.ru/upload/images/2018/01/16/29a2bc.png
- Not working in some skies: http://screenshot.ru/upload/images/2018/01/16/3ef1b2.png

It's really complicated for me to understand whole structure of this boolean, but I know how and where to test it, though.

**Comment by fabiangreffrath** [comment]
Thanks for checking! Is this with or without the 'else' that I removed earlier?

**Comment by JNechaevsky** [comment]
Dang! Just noticed a small flaw in my patch: puffs are not appearing on the floors/ceilings, that's are "longer" than possible hitscan attack distance. Looks likes `dist` is always making straight forward checking for attack range, without accounting of looking direction. Here's the scheme of this issue:

- http://screenshot.ru/upload/images/2018/01/16/dammitbb647.png

**Comment by fabiangreffrath** [comment]
This is because puffs can only appear at the cross section of the shoot trajectory with a line. If the next possible line is farther away than the hitscan attack distance, the bullet puff will never get spawned. AFAIUI you can only mitigate this by extending the hitscan distance ad infinitum, i.e. until you reach the first 1-sided line.

**Comment by fabiangreffrath** [comment]
could you please check again with this patch?

```
--- a/src/doom/p_map.c
+++ b/src/doom/p_map.c
@@ -1081,6 +1081,26 @@ boolean PTR_ShootTraverse (intercept_t* in)
            }
        }

+       // [crispy] fix the pullet puff's z-coordinate if it is outside its
+       // spawning sector's floor or ceiling, respectively
+       {
+               const int lineside = P_PointOnLineSide(x, y, li);
+               int side;
+
+               if ((side = li->sidenum[lineside]) != NO_INDEX)
+               {
+                       const sector_t *const sector = sides[side].sector;
+
+                       if (z < sector->floorheight || z > sector->ceilingheight)
+                       {
+                               z = BETWEEN(sector->floorheight, sector->ceilingheight, z);
+                               frac = FixedDiv(z - shootz, FixedMul(aimslope, attackrange));
+                               x = trace.x + FixedMul (trace.dx, frac);
+                               y = trace.y + FixedMul (trace.dy, frac);
+                       }
+               }
+       }
+
        // [crispy] update laser spot position and return
        if (la_damage == INT_MIN)
        {

```

**Comment by fabiangreffrath** [comment]
> If I'm not mistaking, this addition from Doom Retro's code fixed this issue for me, in my implementation:

Unfortunately, I cannot simply `return false` from the function and omit the spawning of the bullet puff. For demo compatibility, a bullet puff *must* get spawned, but in this specific case (`sector->ceilingpic == skyflatnum`) I could possibly allow the puff to slip along the ceiling so it is probably less noticable.

**Comment by fabiangreffrath** [comment]
PS: Patch updated, in case you would like to check again. ;)

**Comment by JNechaevsky** [comment]
Unbelievable! =O It's working perfect now, last issue on E1M8 ([screenshot](http://screenshot.ru/upload/images/2018/01/18/bandicam2018-01-1811-29-17-765780fe.png)) fixed, in all other placed like frontsectors/backsectors, sky hacks and floors, sharp angles it's working fine already.

But more of that, all three demos from Doom II and all four from Doom I are working perfectly, without any desyncs.

**Comment by fabiangreffrath** [comment]
![40427963](https://user-images.githubusercontent.com/6449237/35088974-ce4ac51a-fc35-11e7-8019-a6a3fcb7c176.jpg)

> But more of that, all three demos from Doom II and all four from Doom I are working perfectly, without any desyncs.

That was part of the plan, but thanks for the confirmation. ;)

**Comment by JNechaevsky** [comment]
Actually, in my last reply, I was honestly going to ask if you know some Higher Magic, but decided to don't do this this because of unnecessary humor. But now there is no need to ask - I know it for sure!

Gime just 10 minutes more, I'll make a more extensive test now by firing ±100,000 bullets in all possible places: more front/back sectors, shooting through the sectors with different heights, different look heights and so on.

**Comment by JNechaevsky** [comment]
Phew, done!
![](http://screenshot.ru/upload/images/2018/01/18/hqdefault31347.jpg)
_(this pic is from "Treasure Island" cartoon, and [this moment](https://youtu.be/Ikit3VEsc-g?t=134) kinda reminded to me myself while testing)_

Noticed just two places:

1. MAP13 ([screenshot](http://screenshot.ru/upload/images/2018/01/18/145b33.png)): puffs are not appearing at all in these small holes. I belive it's not a patch related - sector height is too small (8 map units) to puffs can appear. According to the `info.c`, puff have a height of 16 units: https://github.com/fabiangreffrath/crispy-doom/blob/master/src/doom/info.c#L2188

2. MAP13 again ([screenshot](http://screenshot.ru/upload/images/2018/01/18/247b44.png)): there is small and odd chance that puff may appear on that... Hmm. It seems to be sky ceiling, not sky hack wall. But the looking angle and position must be very exact to see this, so it barely require any attention. Or maybe this is the edge of sky ceiling + sky hack wall, which is not safe to touch for demo compatibility?

**Comment by fabiangreffrath** [comment]
> puffs are not appearing at all in these small holes. I belive it's not a patch related - sector height is too small (8 map units) to puffs can appear.

Hm, is there a restriction that objects must fit into the sector that they get spawned into? I don't think so. Maybe the puff is simply not visible because it gets moved behind the surrounding walls due to some sprite offsets? Any chance to test this with Choco - or even better with the patch commented out?

> Or maybe this is the edge of sky ceiling

Probably. As I stated earlier, the bullet puff *must* be spawned somewhere. In the case of sky ceilings I merely allow the puff to slip to the line at the edge of the sector, where it is *usually* less visible, but in the screenshot you seem to be shooting at exact that area. ;)

**Comment by JNechaevsky** [comment]
> Hm, is there a restriction that objects must fit into the sector that they get spawned into? I don't think so. 

Yes, my bad, you're right. Probably they are appearing outside of map bounds. I can check both cases:
1. Crispy without your patch.
2. Choco with added pillar near that hole.

> As I stated earlier, the bullet puff must be spawned somewhere.

Aha, got it! I think it will non be notified while normal playing, plus this is absolutelly not critical, so no further actions are suggested.

Will post my results about first case shortly.

**Comment by JNechaevsky** [comment]
Done! My results:

**1.** Crispy 5.1 (recently released) - puffs are not appearing: [screenshot](http://screenshot.ru/upload/images/2018/01/18/1_crispyb8f48.png).

**2.** Choco with slightly modified MAP13 ([screenshot](http://screenshot.ru/upload/images/2018/01/18/2_choco0819f0.png)): puffs are appearing, but - they seems to appear in small black sector BELOW the floor, i.e. they are able to spawn, but since their height is taller than sector height, they are appearing below the floor.

Thus, they are also appearing in Crispy, but because of some calculation of sprite clipping (oh no, not again), they simply not visible below `floorheight` + `viewz` stuff.

[Screenshot 1](http://screenshot.ru/upload/images/2018/01/18/2_choco16bff1.png) and [screenshot 2](http://screenshot.ru/upload/images/2018/01/18/2_choco283d1c.png) (puffs are never getting higher from `floorheight` of small black sector).

**Comment by JNechaevsky** [comment]
As a final testing chord, I have tested this map on my project, since I've made small addition to the sprite clipping code in `R_StoreWallRange` ([commit](https://github.com/JNechaevsky/russian-doom/commit/b5eb9ec633bf3f8ff1b97cb0cd8db711ee204317)). Everything is same as in Crispy and Choco, with both enabled and disabled mouse look, which is certainly a good sign. :)

Roughly saying, puff sprite must be visible if only floor texture is visible. But again, it's a sprite clipping stuff, so better don't even to start thinking about it.

For now I'm suggesting to declare that you have done another great and unique thing, that was never achieved before. **Its totally Awesome!**
